File: src/scripts/Denoiser.ts
import metadata from "@/assets/model/model_metadata.json";
import { InferenceSession, Tensor } from "onnxruntime-react-native";

export class DeepFilterNet {
    private session: InferenceSession | null = null;
    private readonly hopSize: number;
    private readonly fftSize: number;
    private readonly stateInputNames: string[];
    private readonly hasSingleStateTensor: boolean;
    private readonly statesShape: number[];
    private readonly statesLen: number;
    private readonly inputNames: string[];
    private readonly outputNames: string[];

    constructor() {
        this.hopSize = metadata.hop_size;
        this.fftSize = metadata.fft_size;
        this.inputNames = metadata.inputs.map((i) => i.name);
        this.outputNames = metadata.outputs.map((o) => o.name);

        this.stateInputNames = this.inputNames.filter(
            (name) => name !== "input_frame" && name !== "atten_lim_db",
        );
        this.hasSingleStateTensor = this.stateInputNames.includes("states");

        if (this.hasSingleStateTensor) {
            const statesMeta = metadata.inputs.find((i) => i.name === "states");
            if (!statesMeta) throw new Error("Metadata for 'states' input not found.");
            this.statesShape = statesMeta.shape;
            this.statesLen = this.statesShape.reduce((a, b) => a * b, 1);
        } else {
            this.statesShape = [];
            this.statesLen = 0;
        }
    }

    public async loadModel(modelPath: string): Promise<void> {
        try {
            const options: InferenceSession.SessionOptions = {
                executionProviders: ["cpu"],
                graphOptimizationLevel: "all",
                interOpNumThreads: 4, // Optimized for multi-core mobile CPUs
                intraOpNumThreads: 4,
            };
            this.session = await InferenceSession.create(modelPath, options);
            console.log(`Successfully loaded model from ${modelPath}`);
        } catch (e) {
            console.error(`Failed to load model: ${e}`);
            throw e;
        }
    }

    private initStates(): {
        states: Tensor | Tensor[];
        attenLimDb: Tensor;
    } {
        let states: Tensor | Tensor[];

        if (this.hasSingleStateTensor) {
            states = new Tensor("float32", new Float32Array(this.statesLen), this.statesShape);
        } else {
            states = [];
            const stateInputsMeta = metadata.inputs.filter((i) =>
                this.stateInputNames.includes(i.name),
            );
            for (const meta of stateInputsMeta) {
                const size = meta.shape.reduce((a, b) => a * b, 1);
                states.push(
                    new Tensor("float32", new Float32Array(size), meta.shape),
                );
            }
        }

        const attenLimDb = new Tensor("float32", new Float32Array([0.0]), [1]);
        return { states, attenLimDb };
    }

    /**
     * Denoise audio feed(s).
     */
    public async denoise(
        audio: Float32Array | Float32Array[],
        onProgress?: ((p: number) => void) | ((p: number, i: number) => void),
    ): Promise<Float32Array | Float32Array[]> {
        if (!this.session) {
            throw new Error("Model not loaded. Please call loadModel() first.");
        }

        const isArray = Array.isArray(audio);
        const feeds = isArray ? audio : [audio];

        // Validate amplitude range for each feed
        for (const feed of feeds) {
            let maxAmplitude = 0;
            for (let i = 0; i < feed.length; i++) {
                const abs = Math.abs(feed[i]);
                if (abs > maxAmplitude) maxAmplitude = abs;
            }
            if (maxAmplitude > 2.0) {
                console.warn(`Audio amplitude (${maxAmplitude}) exceeds expected [-1.0, 1.0] range. Ensure input is float32 PCM.`);
            }
        }

        const results = await this.runDenoiseLoop(feeds, onProgress);

        return isArray ? results : results[0];
    }

    private async runDenoiseLoop(
        audioFeeds: Float32Array[],
        onProgress?: ((p: number) => void) | ((p: number, i: number) => void),
    ): Promise<Float32Array[]> {
        if (!this.session) throw new Error("Session not initialized");

        const numFeeds = audioFeeds.length;
        const origLen = audioFeeds[0].length;
        
        // ALIGNMENT: Prepend fftSize zeros for delay compensation
        const hopSizePadding = (this.hopSize - (origLen % this.hopSize)) % this.hopSize;
        const totalPadding = this.fftSize + hopSizePadding;
        const paddedLen = origLen + totalPadding;
        const numSteps = Math.ceil(paddedLen / this.hopSize);

        // Prepare resources for all feeds
        const feedStates = audioFeeds.map(() => {
            const { states, attenLimDb } = this.initStates();
            const inputFrameData = new Float32Array(this.hopSize);
            const inputFrame = new Tensor("float32", inputFrameData, [this.hopSize]);
            const paddedAudio = new Float32Array(paddedLen);
            // Note: In case of multi-feed, we assume they all have same length
            // We'll set the actual audio later in a loop or here if they are already known
            return {
                states,
                attenLimDb,
                inputFrameData,
                inputFrame,
                paddedAudio,
                enhancedAudio: new Float32Array(numSteps * this.hopSize),
            };
        });

        // Initialize padded audio buffers
        for (let i = 0; i < numFeeds; i++) {
            feedStates[i].paddedAudio.set(audioFeeds[i], this.fftSize);
        }

        let lastReportedProgress = -1;

        for (let step = 0; step < numSteps; step++) {
            const start = step * this.hopSize;

            // Prepare feeds_array for parallel execution
            // then asynchronously iterate through the feeds_array and run this.session.run with every feed
            const inferencePromises = feedStates.map(async (f) => {
                // Fast data copy into reusable tensor buffer
                f.inputFrameData.set(f.paddedAudio.subarray(start, start + this.hopSize));

                const feeds: Record<string, Tensor> = {
                    input_frame: f.inputFrame
                };

                if (this.inputNames.includes("atten_lim_db")) {
                    feeds["atten_lim_db"] = f.attenLimDb;
                }

                if (this.hasSingleStateTensor) {
                    feeds["states"] = f.states as Tensor;
                } else {
                    const stateTensors = f.states as Tensor[];
                    for (let j = 0; j < stateTensors.length; j++) {
                        feeds[this.stateInputNames[j]] = stateTensors[j];
                    }
                }

                const results = await this.session!.run(feeds);

                // Update output buffer and states
                const outFrameRaw = results[this.outputNames[0]].data as Float32Array;
                f.enhancedAudio.set(outFrameRaw, start);

                if (this.hasSingleStateTensor) {
                    const nextStateRaw = results[this.outputNames[1]].data as Float32Array;
                    f.states = new Tensor("float32", nextStateRaw, this.statesShape);
                } else {
                    const newStates: Tensor[] = [];
                    for (let j = 0; j < this.stateInputNames.length; j++) {
                        const nextMultiStateRaw = results[this.outputNames[j + 1]].data as Float32Array;
                        const originalShape = (f.states as Tensor[])[j].dims;
                        newStates.push(new Tensor("float32", nextMultiStateRaw, originalShape));
                    }
                    f.states = newStates;
                }
            });

            await Promise.all(inferencePromises);

            // Progress reporting
            if (onProgress) {
                const progress = Math.round(((step + 1) / numSteps) * 100);
                if (progress > lastReportedProgress) {
                    if (numFeeds > 1) {
                        for (let i = 0; i < numFeeds; i++) {
                            (onProgress as (p: number, i: number) => void)(progress, i);
                        }
                    } else {
                        (onProgress as (p: number) => void)(progress);
                    }
                    lastReportedProgress = progress;
                }
            }
        }

        // Return trimmed denoised buffers
        return feedStates.map(f => f.enhancedAudio.slice(this.fftSize, origLen + this.fftSize));
    }
}

File: src/scripts/denoiser.py
import onnxruntime as ort
import numpy as np
import torch
import torch.nn.functional as F
import torchaudio
from tqdm import tqdm


class OnnxDF:
    """
    Wrapper for a DeepFilterNet ONNX model.
    """
    def __init__(self, model_path: str, device: str = "cpu"):
        self.device = device
        # 1. Load the ONNX runtime session
        providers = ["CPUExecutionProvider"]
        if device == "cuda" and ort.get_device() == 'GPU':
            providers.insert(0, "CUDAExecutionProvider")
        # IDk why it is
        sess_options = ort.SessionOptions()
        sess_options.graph_optimization_level = (
        ort.GraphOptimizationLevel.ORT_ENABLE_EXTENDED
        )
        
        sess_options.intra_op_num_threads = 1
        sess_options.inter_op_num_threads = 1
        sess_options.execution_mode = ort.ExecutionMode.ORT_SEQUENTIAL
        ###-----
        self.sess = ort.InferenceSession(model_path, sess_options, providers=providers)
        print(f"Successfully loaded model from {model_path}")
        print(f"Using provider: {self.sess.get_providers()[0]}")

        # 2. Get model input and output names and shapes from the model metadata.
        self.input_names = [i.name for i in self.sess.get_inputs()]
        self.output_names = [o.name for o in self.sess.get_outputs()]
        print("\nModel Inputs:")
        for i in self.sess.get_inputs():
            print(f"- Name: {i.name}, Shape: {i.shape}, Type: {i.type}")

        print("\nModel Outputs:")
        for o in self.sess.get_outputs():
            print(f"- Name: {o.name}, Shape: {o.shape}, Type: {o.type}")

        # Based on torchDF, we can find the model properties
        self.hop_size = 480 #512
        self.fft_size = 960
        self.frame_size = self.hop_size
        
        # Identify state inputs. They are any inputs that are not the main audio frame or atten_lim_db.
        self.state_input_names = [i.name for i in self.sess.get_inputs() if i.name not in ("input_frame", "atten_lim_db")]
        self.has_single_state_tensor = "states" in self.state_input_names

        if self.has_single_state_tensor:
            # Get the shape of the single 'states' tensor
            states_shape = [s.shape for s in self.sess.get_inputs() if s.name == "states"][0]
            self.states_len = states_shape[0]

    def init_states(self):
        """Initializes the recurrent states of the model."""
        if self.has_single_state_tensor:
            # The model expects a single flattened state tensor.
            states = np.zeros((self.states_len,), dtype=np.float32)
        else:
            # The model expects multiple state tensors.
            states = []
            for i in self.sess.get_inputs():
                if i.name in self.state_input_names:
                    states.append(np.zeros(i.shape, dtype=np.float32))
        atten_lim_db = np.array([0.0], dtype=np.float32)  # Default value
        return states, atten_lim_db

    def __call__(self, audio: torch.Tensor, sr: int):
        """
        Denoises a single-channel audio tensor.
 
        Args:
            audio (torch.Tensor): A single-channel audio tensor of shape [1, T].
            sr (int): The sample rate of the audio.
 
        Returns:
            torch.Tensor: The denoised audio tensor of shape [1, T].
        """
        if sr != 48000:
            print(f"Resampling audio from {sr}Hz to 48000Hz.")
            resampler = torchaudio.transforms.Resample(orig_freq=sr, new_freq=48000)
            audio = resampler(audio)
        if audio.shape[0] > 1:
            print("Warning: Audio has multiple channels, converting to mono.")
            audio = audio.mean(dim=0, keepdim=True)
        if audio.ndim == 1:
            audio = audio.unsqueeze(0)

        # Initialize states for each new audio file
        states, atten_lim_db = self.init_states()

        # Pad audio to be divisible by hop_size
        # This is to ensure we process all samples
        orig_len = audio.shape[-1]
        hop_size_divisible_padding_size = (self.hop_size - orig_len % self.hop_size) % self.hop_size
        # This padding is based on the original implementation to align with the STFT processing
        padded_audio = F.pad(audio, (0, self.fft_size + hop_size_divisible_padding_size))
        audio_chunks = torch.split(padded_audio.squeeze(0), self.hop_size)
        
        enhanced_frames = []
        
        print("Denoising audio...")
        for frame in tqdm(audio_chunks):
            # Prepare input feed for ONNX session
            input_feed = {"input_frame": frame.numpy()}
            if "atten_lim_db" in self.input_names:
                input_feed["atten_lim_db"] = atten_lim_db

            if self.has_single_state_tensor:
                input_feed["states"] = states
            else:
                for name, state_val in zip(self.state_input_names, states):
                    input_feed[name] = state_val

            # Run inference
            outputs = self.sess.run(self.output_names, input_feed)
            enhanced_frame = outputs[0]

            # Update states for the next frame
            if self.has_single_state_tensor:
                states = outputs[1]  # new_states is the second output
            else:
                # The new states are the outputs after the enhanced frame
                states = outputs[1:len(self.state_input_names)+1]

            enhanced_frames.append(torch.from_numpy(enhanced_frame))

        enhanced_audio = torch.cat(enhanced_frames).unsqueeze(0)
        # Trim the audio to the original length, accounting for the model's delay
        delay = self.fft_size - self.hop_size
        enhanced_audio = enhanced_audio[:, delay : orig_len + delay]
        return enhanced_audio



def main(args):
    """
    Main function to load model, process audio, and save the result.
    """
    try:
        # Load audio file
        noisy_audio, sr = torchaudio.load(args.input_path)
        print(f"Loaded audio from {args.input_path} with sample rate {sr}.")
    except Exception as e:
        print(f"Error loading audio file: {e}")
        return

    # Initialize the denoiser
    denoiser = OnnxDF(args.model_path)

    # Denoise the audio
    enhanced_audio = denoiser(noisy_audio, sr)

    # Save the enhanced audio
    try:
        torchaudio.save(
            args.output_path,
            enhanced_audio,
            48000,
            encoding="PCM_S",
            bits_per_sample=16,
        )
        print(f"Saved enhanced audio to {args.output_path}")
    except Exception as e:
        print(f"Error saving audio file: {e}")


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description="Denoise an audio file using a DeepFilterNet ONNX model.")
    parser.add_argument(
        "-m", "--model_path", 
        type=str, 
        default="./denoiser_model.onnx", 
        help="Path to the denoiser ONNX/ORT model file."
    )
    parser.add_argument(
        "-i", "--input_path", 
        type=str, 
        required=True, 
        help="Path to the noisy input audio file (WAV format)."
    )
    parser.add_argument(
        "-o", "--output_path", 
        type=str, 
        default="enhanced_audio.wav", 
        help="Path to save the enhanced output audio file."
    )
    
    cli_args = parser.parse_args()
    main(cli_args)
